lab2/lab2.md

```diff
@@ -1,30 +1,624 @@
+
+The simple in-order CPU design is based closely on the CPU model in Patterson and Hennessey's Computer Organization and Design.
+
+
+This is the first time we have used this set of assignments, so there may be mistakes.
+To offset this, we will be offering a variety of ways to get extra credit.
+See [the extra credit page](../extra-credit.md) for an up-to-date list of ways to get extra credit.
+
+
+
+# Part I: Implement the Branch-Control
+
+**The test for this part is `dinocpu.BranchControlTesterLab2`.**
+
+In this part you will be implementing the branch-control component in the CPU design. The branch-control, controls whether or not branches are taken.
+
+It takes four inputs: `branch`, `funct3`, `inputx`, `inputy` and generates one output `taken`.
+
+```
+ 
+branch: true if we are looking at a branch
+funct3: the middle three bits of the instruction (12-14). Specifies the type of branch
+inputx: first value (e.g., reg1)
+inputy: second value (e.g., reg2)
+taken: true if the branch is taken.
+```
+
+The following table details the `funct3` input and which type of branching condition to test:
+
+|      |     |
+|------|-----|
+| 000 | beq  |
+| 001 | bne  |
+| 100 | blt  |
+| 101 | bge  |
+| 110 | bltu |
+| 111 | bgeu |
+
+
+You must take the RISC-V ISA specification and implement the proper control to choose the right type of branch test and correctly set or reset the `taken` output if the branch test passes or fails respectively. 
+You can find the specification in the following places:
+
+
+- [on the RISC-V website](https://riscv.org/specifications/)
+
+|31                 25|24  20|19 15|14     12|11                 7|6       0|
+|--------|------------|------|-----|---------|-----------|--------|---------|--------|
+|imm[12] | imm[10:5]  | rs2  | rs1 | funct3  | imm[4:1]  | imm[11]|  opcode | B-type |
+|--------|------------|------|-----|---------|-----------|--------|---------|--------|
+|     imm[12|10:5]]   | rs2  | rs1 |   000   |     imm[4:1|11]    | 1100011 |  BEQ   |
+|     imm[12|10:5]]   | rs2  | rs1 |   001   |     imm[4:1|11]    | 1100011 |  BNE   |
+|     imm[12|10:5]]   | rs2  | rs1 |   100   |     imm[4:1|11]    | 1100011 |  BLT   |
+|     imm[12|10:5]]   | rs2  | rs1 |   101   |     imm[4:1|11]    | 1100011 |  BGE   |
+|     imm[12|10:5]]   | rs2  | rs1 |   110   |     imm[4:1|11]    | 1100011 |  BLTU  |
+|     imm[12|10:5]]   | rs2  | rs1 |   111   |     imm[4:1|11]    | 1100011 |  BGEU  |
+
+
+This table is from the RISC-V User-level ISA Specification v2.2, page 104.
+You can find the same information in Chapter 2 of the Specification, Chapter 2 of the RISC-V reader, or in the front of the Computer Organization and Design book.
+
+
+Given these inputs, you must generate the correct output on the taken wire.
+The template code from `src/main/scala/components/branch-control.scala` is shown below.
+You will fill in where it says *Your code goes here*.
```

Comment:
Add something here about how they can use `<`, `>`, `===`, and `=/=`, as well as .toUInt and .toSInt